builtins/web/fetch/headers.cpp

```diff
@@ -923,12 +1080,181 @@ unique_ptr<host_api::HttpHeaders> Headers::handle_clone(JSContext* cx, HandleObj
+      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersList)).toPrivate());
+  MOZ_ASSERT(headers_list);
+
+  ensure_updated_sort_list(headers_list, headers_sort_list);
+  MOZ_RELEASE_ASSERT(index < headers_sort_list->size());
+
+  return &headers_list->at(headers_sort_list->at(index));
+}
+
+std::optional<size_t> Headers::lookup(JSContext *cx, HandleObject self, string_view key) {
+  MOZ_ASSERT(is_instance(self));
+  const HeadersList *headers_list = static_cast<Headers::HeadersList *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersList)).toPrivate());
+  MOZ_ASSERT(headers_list);
+  std::vector<size_t> *headers_sort_list = static_cast<std::vector<size_t> *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Headers::Slots::HeadersSortList)).toPrivate());
+  MOZ_ASSERT(headers_sort_list);
+
+  ensure_updated_sort_list(headers_list, headers_sort_list);
+
+  // Now we know its sorted, we can binary search.
+  auto it = std::lower_bound(headers_sort_list->begin(), headers_sort_list->end(), key,
+                             HeadersSortListLookupCompare(headers_list));
+  if (it == headers_sort_list->end() ||
+      header_compare(std::get<0>(headers_list->at(*it)), key) != Ordering::Equal) {
+    return std::nullopt;
+  }
+  return it - headers_sort_list->begin();
+}
+
+bool HeadersIterator::next(JSContext *cx, unsigned argc, Value *vp) {
+  METHOD_HEADER(0)
+  JS::RootedObject headers(cx, &JS::GetReservedSlot(self, Slots::Headers).toObject());
+
+  Headers::HeadersList *list = Headers::get_list(cx, headers);
+
+  size_t index = JS::GetReservedSlot(self, Slots::Cursor).toInt32();
+  size_t len = list->size();
+  uint8_t type = static_cast<uint8_t>(JS::GetReservedSlot(self, Slots::Type).toInt32());
+
+  JS::RootedObject result(cx, JS_NewPlainObject(cx));
+  if (!result)
+    return false;
+
+  if (index >= len) {
+    JS_DefineProperty(cx, result, "done", JS::TrueHandleValue, JSPROP_ENUMERATE);
+    JS_DefineProperty(cx, result, "value", JS::UndefinedHandleValue, JSPROP_ENUMERATE);
+
+    args.rval().setObject(*result);
+    return true;
+  }
+
+  JS_DefineProperty(cx, result, "done", JS::FalseHandleValue, JSPROP_ENUMERATE);
+
+  JS::RootedValue key_val(cx);
+  JS::RootedValue val_val(cx);
+
+  if (type != ITER_TYPE_VALUES) {
+    const host_api::HostString *key = &std::get<0>(*Headers::get_index(cx, headers, index));
+    size_t len = key->len;
+    JS::Latin1Char *chars = reinterpret_cast<JS::Latin1Char *>(malloc(len));
+    for (int i = 0; i < len; ++i) {
+      const unsigned char ch = key->ptr[i];
+      // headers should already be validated by here
+      MOZ_ASSERT(ch <= 127 && VALID_NAME_CHARS[ch]);
+      // we store header keys with casing, so getter itself lowercases
+      if (ch >= 'A' && ch <= 'Z') {
+        chars[i] = ch - 'A' + 'a';
+      } else {
+        chars[i] = ch;
+      }
+    }
+    key_val = JS::StringValue(JS_NewLatin1String(cx, JS::UniqueLatin1Chars(chars), len));
```

Comment:
This needs an OOM check for the `JS_NewLatin1String` before creating the `StringValue`

Reply:
`JS_NewLatin1String` takes ownership of the unique chars, and so there is no memory allocation or failure associated.